### ISSUE TITLE
Lock Notes via PIN/Fingerprint 

### ISSUE THREAD

**Issue by PhilKes** [PhilKes/NotallyX#30]
I had a look at some Google Play Store reviews for Notally, and this was mentioned multiple times.

It should be possible to lock viewing the notes via PIN/Fingerprint, similar to e.g. in WhatsApp

**Comment by ippocratis** [comment]
This together with encrypted backups will enhance privacy/security of the app greatly.

**Comment by PhilKes** [comment]
I'm not that experience with Android Apps' security, but right now the database with the notes is stored in the app's internal storage which can only be accessed if you have rooted your device. So I guess for now its sufficient to just provide a `BiometricPrompt` that the user has to authenticate with and then show the `MainActivity`?

In #18 we can then encrypt the `.zip` files with a user set password

**Comment by ippocratis** [comment]
Encrypting the exported backuos and biometric unlock is surely essential .

OK not sure if you imply that biometric/password app lock also means db encryption but i think it alao makes sense to  encrypt the sqlite db with something like SQLCipher and decrypt it upon app unlock.

No idea how hard or even doable this might be.

For the locking frequency I think all options make sense.
Relocking when device is locked.
Relocking when the app exits from recents.
Relocking after a timeout.

If I had to choose one , probably is choose relock upon device lock.

**Comment by PhilKes** [comment]
> Encrypting the exported backups and biometric unlock is surely essential .
> 
> I'm not sure if you imply that biometric/password app lock also means db encryption but i think it also makes sense to encrypt the sqlite db with something like SQLCipher and decrypt it upon app unlock.
> 
> No idea how hard or even doable this might be.
> 
> For the locking frequency I think all options make sense. Relocking when device is locked. Relocking when the app exits from recents. Relocking after a timeout.
> 
> If I had to choose one , probably is relock upon device lock.

I'm currently looking into how to encrypt/decrypt the database with sqlcipher and with the `BiometricPrompt.AuthenticationResult.cryptoObject` but I'm having a really hard time. I also dont find any reference implementations for sqlcipher + BiometricPrompt with cryptoObject

**Comment by ippocratis** [comment]
I wouldn't burn out to features that require to much effort.

Just stick to app lock and encrypted backups and leave database encryption for later.

App sandbox provide a first stage of security anyway.

**Comment by CodeNomadHQ** [comment]
I’d love to know how far you’ve gotten with your project. Have you added the necessary dependencies for SQLCipher and BiometricPrompt in your build.gradle yet? If you’re running into any challenges, I’d be happy to help out a bit.

**Comment by PhilKes** [comment]
> I’d love to know how far you’ve gotten with your project. Have you added the necessary dependencies for SQLCipher and BiometricPrompt in your build.gradle yet? If you’re running into any challenges, I’d be happy to help out a bit.

I just pushed my PR, which does **not** include encrypting the database yet.
The SQLCipher simply adds a Helper to the Room database, but it wants a passphrase, while the `BiometricPrompt.AuthenticationResult` gives me a Cipher with IV, and I dont yet understand what I am supposed to do with it.
Am I supposed to encrypt/decrypt a random passphrase that I store in `SharedPreferences` or something?

**Comment by PhilKes** [comment]
I just pushed some new changes to https://github.com/PhilKes/NotallyX/pull/42
I managed to encrypt the database now too.

1. Enable Lock in Settings
2. Authenticate via Biometric/PIN prompt (with initialized `Cipher` for encryption)
3. On successfull authentication the `Cipher.iv` is stored in the `Preferences`
4. A `SecretKey` in the AndroidKeyStore is created, as a `passphrase` for `SQLCipher`
5. Database gets encrypted with the `passphrase` and `SQLCipherUtils`
6. The `passphrase` is encrypted with the `Cipher` and stored in the `Preferences`
7. From now on the `RoomDatabase` is built with SQLCipher's `SupportFactory`, by decrypting the `passphrase` from the preferences with the `iv`

I dont know if this is really how its supposed to be done, as I said before, I found no example of encrypting a `RoomDatabase` via `BiometricPrompt`.

What I do not understand is, when the lock is setup, the database encrypted, am I supposed to use the `Cipher` I get from the `BiometricPrompt.AuthenticationResult`? I have to initialize that Cipher for Decryption with the `iv` in the preference before showing the `BiometricPrompt`, so I dont get what I am supposed to do with the decryption `Cipher`
